# State Farm Data Scientist Opening Pre-employment Assessment

Rezoan Ahmed Shuvro
email: rezoanshuvro@gmail.com

## Initial data exploration

In [1]:
# Importing the libraries and loading the .csv file in a dataframe
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler
import warnings
warnings.filterwarnings('ignore')

In [2]:
# loading the dataset
dataset = pd.read_csv('exercise_01_train.csv')
test =pd.read_csv('exercise_01_test.csv')

In [3]:
# head of the data set
dataset.head(5)

x0         x1         x2         x3         x4         x5        x6  \
0  10.142889 -15.675620   3.583176 -22.397489  27.221894 -34.110924 -0.072829   
1 -52.214630   5.847135 -10.902843 -14.132351  20.588574  36.107322  0.115023   
2  67.718500   2.064334  12.394186 -18.667102  47.465504 -50.373658  0.253707   
3 -28.003111   8.565128  -8.592092   5.918960  -3.224154  78.315783 -0.879845   
4  80.703016  30.736353 -30.101857 -21.201140 -91.946233 -47.469246 -0.646831   

         x7        x8        x9 ...       x91        x92     x93       x94  \
0 -0.544444  0.997601 -2.691778 ...  1.916575   5.240820  euorpe  2.431170   
1  0.276093 -0.699168 -0.972708 ...  0.370941  -3.794542    asia  2.592326   
2  1.068968  2.939713  2.691218 ...  1.449817  12.470532    asia  7.143821   
3  1.176889 -2.414752  0.589646 ... -3.274733   3.484450    asia -4.998195   
4 -0.578398  0.980849 -1.426112 ... -0.644261   4.082783    asia -0.012556   

         x95        x96        x97       x98        x99  y  
0   0.454074 -18.572032 -14.291524  0.178579  18.110170  0  
1  31.921833   3.317139  10.037003 -1.930870  -3.486898  0  
2   9.401490 -10.604968   7.643215 -0.842198 -79.358236  0  
3 -20.312810  14.818524  -9.180674  1.356972  14.475681  0  
4 -29.334324   1.734433 -12.262072 -0.043228 -19.003881  0  

[5 rows x 101 columns]

In [4]:
# Tail of the dataset to check the size is okay
dataset.tail(2)

x0         x1        x2         x3         x4         x5  \
39998  50.168318  -4.272643  2.409248 -11.697615  39.234827  31.353302   
39999  -8.653274  10.572796  1.377445 -21.472814 -42.686853  28.893360   

             x6        x7        x8        x9 ...       x91        x92  \
39998  1.416008  1.825775  2.027886 -3.753114 ... -1.411384 -17.587621   
39999  3.379456 -1.241659 -0.040278  0.612898 ...  7.622624  -6.473851   

           x93       x94        x95        x96       x97       x98        x99  \
39998  america  6.278226 -18.743967  -8.067506  5.258203 -2.623772 -15.550075   
39999     asia  0.055730  -6.506186  12.434701 -6.001283 -5.340633  18.276723   

       y  
39998  0  
39999  1  

[2 rows x 101 columns]

In [5]:
#columns of the dataset
dataset.columns

Index(['x0', 'x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7', 'x8', 'x9',
       ...
       'x91', 'x92', 'x93', 'x94', 'x95', 'x96', 'x97', 'x98', 'x99', 'y'],
      dtype='object', length=101)

In [6]:
# output variable. Output is imbalanced. 80% samples represent 0 class and 20% samples represent 1 class
dataset.y.value_counts('normalize=True')

0    0.798825
1    0.201175
Name: y, dtype: float64

In [7]:
# shape of the dataset
dataset.shape

(40000, 101)

In [8]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Columns: 101 entries, x0 to y
dtypes: float64(94), int64(1), object(6)
memory usage: 30.8+ MB


In [9]:
dataset.describe()

x0            x1            x2            x3            x4  \
count  39986.000000  39990.000000  39994.000000  39990.000000  39994.000000   
mean       8.259955     -3.249786      1.030666     -0.747566      0.283820   
std       38.374182     15.171131     24.732185     15.225730     42.240018   
min     -140.780478    -64.493908   -105.388182    -63.804916   -158.195975   
25%      -17.800204    -13.458580    -15.565461    -11.078276    -28.246509   
50%        8.354662     -3.386601      1.132995     -0.714888      0.292788   
75%       33.829780      6.881661     17.677615      9.552404     28.719663   
max      177.399176     62.906822     99.394915     59.338352    179.342581   

                 x5            x6            x7            x8            x9  \
count  39990.000000  39993.000000  39987.000000  39994.000000  39993.000000   
mean      -1.773510     -0.000232     -0.016107     -0.651093     -0.014688   
std       42.124100      1.065955      3.382644      2.947472      1.906496   
min     -169.237259     -4.133490    -12.966970    -12.037625     -7.446200   
25%      -30.391354     -0.723098     -2.299081     -2.628856     -1.299759   
50%       -1.753365      0.001105     -0.003556     -0.659223     -0.028170   
75%       26.844781      0.715844      2.259972      1.322101      1.263469   
max      170.894497      5.311653     16.619445     14.994937      7.300186   

           ...                x90           x91           x92           x94  \
count      ...       39993.000000  39998.000000  39994.000000  39989.000000   
mean       ...         -14.274809      0.011390      0.003948     -0.050510   
std        ...         154.038206      3.311041      8.763944      4.979690   
min        ...        -674.004008    -12.807938    -38.121111    -21.578977   
25%        ...        -116.645845     -2.218739     -5.925508     -3.435180   
50%        ...         -11.471306     -0.006726      0.009306     -0.037111   
75%        ...          90.101751      2.238996      5.909011      3.299108   
max        ...         603.911528     14.982369     35.785334     20.983463   

                x95           x96           x97           x98           x99  \
count  39993.000000  39985.000000  39993.000000  39995.000000  39987.000000   
mean      -0.007572     -0.629241     -1.986671      0.036482      1.486887   
std       19.238210     16.915222     14.375663      5.633052     36.926796   
min      -87.669573    -77.010252    -57.709983    -23.588876   -154.559512   
25%      -12.895717    -11.948902    -11.686033     -3.770599    -23.559519   
50%        0.124945     -0.481374     -2.026059      0.041838      1.465346   
75%       12.988509     10.793171      7.611660      3.840100     26.548474   
max       78.785164     70.182932     60.481075     22.759016    143.126382   

                  y  
count  40000.000000  
mean       0.201175  
std        0.400884  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000  

[8 rows x 95 columns]

First checking the classification using Logistic regression without any tuning, removing all the missing values and without any preprocessing. This will be my base accuracy which i will improve subsequently.

In [10]:
dataset[dataset==np.inf]=np.nan
dataset.fillna(dataset.mean(),inplace=True)
X =dataset.drop(['x34','x35','x68','x93','x41','x45','y'],axis=1)
y = dataset['y']
#scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X = sc.fit_transform(X)

#train test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)

#training
from sklearn.linear_model import LogisticRegression
logmodel = LogisticRegression()
logmodel.fit(X_train,y_train)

# prediction using model
from sklearn.metrics import accuracy_score
prediction = logmodel.predict(X_test)
#evaluation(Accuracy)
from sklearn.metrics import accuracy_score

print("Accuracy:",accuracy_score(prediction,y_test))
#evaluation
from sklearn.metrics import classification_report
print(classification_report(y_test,prediction))

Accuracy: 0.8905
              precision    recall  f1-score   support

           0       0.91      0.96      0.93      6392
           1       0.80      0.61      0.69      1608

   micro avg       0.89      0.89      0.89      8000
   macro avg       0.85      0.79      0.81      8000
weighted avg       0.89      0.89      0.88      8000



Due to the imbalance in class, accuracy is high but the recall and f1-score is low. For this dataset, our metric should be recall/f1-score than accuracy. Base f1-score is 0.69


In [11]:
# beta co-efficients
logmodel.coef_

array([[ 2.01456013e-03,  4.30316858e-01,  1.35901214e-01,
        -2.66219022e-01,  6.02327509e-02, -7.13458380e-02,
        -5.40569174e-03, -1.48344131e-02, -4.34590294e-01,
         2.57120714e-02, -5.79191196e-01, -3.20165691e-05,
        -7.36871570e-01,  7.68410647e-03, -3.96408828e-02,
         1.50687548e-02,  8.68193588e-03, -1.41536502e-02,
        -1.21439551e-01,  2.28686513e-01, -2.96120871e-01,
        -4.17653433e-01, -3.89955224e-02,  5.14367059e-04,
        -1.59067113e-02,  4.05608614e-02, -8.47889280e-03,
         2.81940112e-01,  2.78561074e-03, -2.36600900e-02,
        -1.95050545e-02, -2.27600029e-02,  2.54537193e-04,
         1.24398258e-01,  1.74449092e-02, -8.21442885e-01,
         4.97526654e-02,  1.45646394e-02,  3.90238587e-01,
         6.71444858e-01, -3.43083103e-01,  1.68317792e+00,
         2.81398018e-03, -1.48306085e-01, -1.01843447e-02,
         1.62253412e-01, -8.06722032e-01,  4.58456552e-01,
         3.30287686e-02, -3.10942916e-01,  9.28351822e-0

##  Data Cleaning

In [12]:
dataset.head()

x0         x1         x2         x3         x4         x5        x6  \
0  10.142889 -15.675620   3.583176 -22.397489  27.221894 -34.110924 -0.072829   
1 -52.214630   5.847135 -10.902843 -14.132351  20.588574  36.107322  0.115023   
2  67.718500   2.064334  12.394186 -18.667102  47.465504 -50.373658  0.253707   
3 -28.003111   8.565128  -8.592092   5.918960  -3.224154  78.315783 -0.879845   
4  80.703016  30.736353 -30.101857 -21.201140 -91.946233 -47.469246 -0.646831   

         x7        x8        x9 ...       x91        x92     x93       x94  \
0 -0.544444  0.997601 -2.691778 ...  1.916575   5.240820  euorpe  2.431170   
1  0.276093 -0.699168 -0.972708 ...  0.370941  -3.794542    asia  2.592326   
2  1.068968  2.939713  2.691218 ...  1.449817  12.470532    asia  7.143821   
3  1.176889 -2.414752  0.589646 ... -3.274733   3.484450    asia -4.998195   
4 -0.578398  0.980849 -1.426112 ... -0.644261   4.082783    asia -0.012556   

         x95        x96        x97       x98        x99  y  
0   0.454074 -18.572032 -14.291524  0.178579  18.110170  0  
1  31.921833   3.317139  10.037003 -1.930870  -3.486898  0  
2   9.401490 -10.604968   7.643215 -0.842198 -79.358236  0  
3 -20.312810  14.818524  -9.180674  1.356972  14.475681  0  
4 -29.334324   1.734433 -12.262072 -0.043228 -19.003881  0  

[5 rows x 101 columns]

In [13]:
# Fill the missing values in each column using the mean of than column
dataset.fillna(dataset.mean(),inplace=True)
test.fillna(test.mean(),inplace=True)



In [14]:
# Fill in the missing values in the categorical columns using the largest value
dataset= dataset.apply(lambda x:x.fillna(x.value_counts().index[0]))
test = test.apply(lambda x:x.fillna(x.value_counts().index[0]))
test.isnull().sum().sum()

0

In [15]:
# check if there is any missing value lest after imputing the missing values
print('missing value in dataset:',dataset.isnull().sum().sum())
print('missing value in testset:',test.isnull().sum().sum())

missing value in dataset: 0
missing value in testset: 0


There are missing values in the dataset but very few in each column. I will fill the missing values with the mean.

In [16]:
# Examine the data types of the DataFrame
dataset.dtypes

x0     float64
x1     float64
x2     float64
x3     float64
x4     float64
x5     float64
x6     float64
x7     float64
x8     float64
x9     float64
x10    float64
x11    float64
x12    float64
x13    float64
x14    float64
x15    float64
x16    float64
x17    float64
x18    float64
x19    float64
x20    float64
x21    float64
x22    float64
x23    float64
x24    float64
x25    float64
x26    float64
x27    float64
x28    float64
x29    float64
        ...   
x71    float64
x72    float64
x73    float64
x74    float64
x75    float64
x76    float64
x77    float64
x78    float64
x79    float64
x80    float64
x81    float64
x82    float64
x83    float64
x84    float64
x85    float64
x86    float64
x87    float64
x88    float64
x89    float64
x90    float64
x91    float64
x92    float64
x93     object
x94    float64
x95    float64
x96    float64
x97    float64
x98    float64
x99    float64
y        int64
Length: 101, dtype: object

In [17]:
# column x35 represents the days.
dataset.groupby('x35').y.value_counts('normalize=True')

x35        y
fri        0    0.666667
           1    0.333333
friday     0    0.823194
           1    0.176806
monday     0    0.553571
           1    0.446429
thur       0    0.836991
           1    0.163009
thurday    0    0.828281
           1    0.171719
tuesday    0    0.652314
           1    0.347686
wed        0    0.792037
           1    0.207963
wednesday  0    0.744526
           1    0.255474
Name: y, dtype: float64

In [18]:
# x34 represents various counts of car
dataset.x34.value_counts()

volkswagon    12579
Toyota        10946
bmw            7304
Honda          5129
tesla          2275
chrystler      1223
nissan          336
ford            165
mercades         32
chevrolet        11
Name: x34, dtype: int64

In [19]:
#date counts
dataset.x35.value_counts()

wed          14844
thurday      13359
wednesday     5891
thur          4374
tuesday        929
friday         526
monday          56
fri             21
Name: x35, dtype: int64

In [20]:
#month counts
dataset.x68.value_counts()

July       11097
Jun         9279
Aug         8155
May         4788
sept.       3466
Apr         1641
Oct          926
Mar          409
Nov          156
Feb           54
Dev           18
January       11
Name: x68, dtype: int64

In [21]:
#continent counts
dataset.x93.value_counts()

asia       35416
america     3136
euorpe      1448
Name: x93, dtype: int64

In [22]:
# convert categorical variables using one-hot encoder

dum_df = pd.get_dummies(data=dataset, columns=['x34','x35','x68','x93'], prefix=['x34_','x35_','x68_','x93_'])

test_df = pd.get_dummies(data=test, columns=['x34','x35','x68','x93'], prefix=['x34_','x35_','x68_','x93_'])

In [23]:
dum_df.head()

x0         x1         x2         x3         x4         x5        x6  \
0  10.142889 -15.675620   3.583176 -22.397489  27.221894 -34.110924 -0.072829   
1 -52.214630   5.847135 -10.902843 -14.132351  20.588574  36.107322  0.115023   
2  67.718500   2.064334  12.394186 -18.667102  47.465504 -50.373658  0.253707   
3 -28.003111   8.565128  -8.592092   5.918960  -3.224154  78.315783 -0.879845   
4  80.703016  30.736353 -30.101857 -21.201140 -91.946233 -47.469246 -0.646831   

         x7        x8        x9     ...       x68__July  x68__Jun  x68__Mar  \
0 -0.544444  0.997601 -2.691778     ...               0         1         0   
1  0.276093 -0.699168 -0.972708     ...               1         0         0   
2  1.068968  2.939713  2.691218     ...               1         0         0   
3  1.176889 -2.414752  0.589646     ...               0         0         0   
4 -0.578398  0.980849 -1.426112     ...               1         0         0   

   x68__May  x68__Nov  x68__Oct  x68__sept.  x93__america  x93__asia  \
0         0         0         0           0             0          0   
1         0         0         0           0             0          1   
2         0         0         0           0             0          1   
3         1         0         0           0             0          1   
4         0         0         0           0             0          1   

   x93__euorpe  
0            1  
1            0  
2            0  
3            0  
4            0  

[5 rows x 130 columns]

In [24]:
# convert currency to float (column x41))

dum_df['x41'] = dum_df['x41'].str.replace(',', '')
dum_df['x41'] = dum_df['x41'].str.replace('$', '')
dum_df.x41.astype(float)

test_df['x41'] = test_df['x41'].str.replace(',', '')
test_df['x41'] = test_df['x41'].str.replace('$', '')
test_df.x41.astype(float)

0      -1073.61
1       1775.77
2        697.23
3       -134.48
4       1195.16
5      -1424.68
6       -617.48
7       1022.59
8       1041.35
9        272.05
10       977.73
11     -2040.44
12      1288.77
13     -1496.64
14      -477.55
15       472.64
16       399.79
17      -301.75
18       663.79
19      -270.93
20      1890.87
21      -269.77
22       783.73
23       639.48
24      -529.72
25      -191.83
26     -1362.41
27      1663.55
28      -543.37
29       694.77
         ...   
9970    -452.28
9971    2106.34
9972     410.84
9973   -1116.36
9974     622.79
9975    -513.71
9976    -551.90
9977   -1431.80
9978   -1000.70
9979    -704.58
9980     287.84
9981     575.40
9982    -602.24
9983      43.54
9984    -134.64
9985    -286.30
9986   -1813.54
9987    -462.39
9988     424.59
9989   -1058.33
9990     694.72
9991    -465.20
9992    1145.35
9993     -94.50
9994     317.62
9995    -197.68
9996    2084.39
9997    -128.47
9998     293.73
9999    1202.18
Name: x41, Length: 10000

In [25]:
# convert % to float (column x41))
dum_df['x45'] = dum_df['x45'].str.replace(',', '')
dum_df['x45'] = dum_df['x45'].str.replace('%', '')
dum_df.x41.astype(float)

test_df['x45'] = dum_df['x45'].str.replace(',', '')
test_df['x45'] = dum_df['x45'].str.replace('%', '')
test_df.x41.astype(float)

0      -1073.61
1       1775.77
2        697.23
3       -134.48
4       1195.16
5      -1424.68
6       -617.48
7       1022.59
8       1041.35
9        272.05
10       977.73
11     -2040.44
12      1288.77
13     -1496.64
14      -477.55
15       472.64
16       399.79
17      -301.75
18       663.79
19      -270.93
20      1890.87
21      -269.77
22       783.73
23       639.48
24      -529.72
25      -191.83
26     -1362.41
27      1663.55
28      -543.37
29       694.77
         ...   
9970    -452.28
9971    2106.34
9972     410.84
9973   -1116.36
9974     622.79
9975    -513.71
9976    -551.90
9977   -1431.80
9978   -1000.70
9979    -704.58
9980     287.84
9981     575.40
9982    -602.24
9983      43.54
9984    -134.64
9985    -286.30
9986   -1813.54
9987    -462.39
9988     424.59
9989   -1058.33
9990     694.72
9991    -465.20
9992    1145.35
9993     -94.50
9994     317.62
9995    -197.68
9996    2084.39
9997    -128.47
9998     293.73
9999    1202.18
Name: x41, Length: 10000

In [26]:
dum_df.y.value_counts(normalize=True)

0    0.798825
1    0.201175
Name: y, dtype: float64

Next, i will do upsampling to balance the dataset

In [27]:
# upsampling

X =dum_df.drop(['y'],axis=1)
y = dum_df['y']

#train test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)

# concatenate our training data back together
X = pd.concat([X_train, y_train], axis=1)

# separate minority and majority classes
zero = X[X.y==0]
one = X[X.y==1]


# upsample minority
from sklearn.utils import resample
one_upsampled = resample(one,
                          replace=True, # sample with replacement
                          n_samples=len(zero), # match number in majority class
                          random_state=101) # reproducible results

# combine majority and upsampled minority
upsampled = pd.concat([zero, one_upsampled])



X =upsampled.drop(['y'],axis=1)
y = upsampled['y']

In [28]:
#check if balanced
upsampled.y.value_counts(normalize=True)

1    0.5
0    0.5
Name: y, dtype: float64

The dataset is now balanced

The dataset has 130 columns. I will do PCA to reduce the dimensionality of the dataset to 10 from 130. 

In [29]:
# PCA

from sklearn.decomposition import PCA


X =upsampled.drop(['y'],axis=1)
y = upsampled['y']
#scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_pca = sc.fit_transform(X)


pca = PCA(n_components=20)
pca.fit(X_pca)
x_pca = pca.transform(X_pca)

Using reduced dimensionality and 20 features i have achieved 0.72 f1-score with logistic regression and 0.96 f1-score using random forest model. Since the f1-score decreased after using PCA, i decided against using PCA. I rollback to the original X,y i created after upsampling  

The dataset is now clean and ready for modeling.

## Modeling

First, i will again implement the logistic regression model to check the f1-score. Recall that the base f1-score without any data cleaning was 0.69

In [30]:
#logistic regression
#train test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)

#training
from sklearn.linear_model import LogisticRegression
logmodel = LogisticRegression()
logmodel.fit(X_train,y_train)

# prediction using model
from sklearn.metrics import accuracy_score
prediction = logmodel.predict(X_test)
#evaluation(Accuracy)
from sklearn.metrics import accuracy_score

print("Accuracy:",accuracy_score(prediction,y_test))
#evaluation
from sklearn.metrics import classification_report
print(classification_report(y_test,prediction))

Accuracy: 0.8308068459657701
              precision    recall  f1-score   support

           0       0.83      0.83      0.83      5119
           1       0.83      0.83      0.83      5106

   micro avg       0.83      0.83      0.83     10225
   macro avg       0.83      0.83      0.83     10225
weighted avg       0.83      0.83      0.83     10225



The f1-score now is 0.83, which is a significant improvement compared to 0.69

### Now i will use the following two algorithms to classify with better accuracy and tune the hyperparameters to improve the accuracy

1. Random forest (using sklearn library)
2. Artificial Neural Netwrok (using Tensorflow with Keras wrapper)

In [31]:
# Random Forest with hyperparameter tuning parameters
# importing modules
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier

#making the instance
model=RandomForestClassifier()

#hyper parameters set
params = {'criterion':['gini','entropy'],
          'n_estimators':[50,150,300],
          'min_samples_leaf':[5,10],
          'min_samples_split':[5,10]}

#Making models with hyper parameters sets
model = GridSearchCV(model, param_grid=params)

#Learning
model.fit(X_train, y_train)

#The best hyper parameters set
print("Best Hyper Parameters:\n",model.best_params_)

#Prediction
prediction=model.predict(X_test)

#importing the metrics module
from sklearn import metrics

#evaluation(Accuracy)
print("Accuracy:",metrics.accuracy_score(prediction,y_test))

#evaluation(Confusion Metrix)
print(classification_report(y_test,prediction))

Best Hyper Parameters:
 {'criterion': 'entropy', 'min_samples_leaf': 5, 'min_samples_split': 5, 'n_estimators': 300}
Accuracy: 0.9908068459657702
              precision    recall  f1-score   support

           0       0.99      0.99      0.99      5119
           1       0.99      0.99      0.99      5106

   micro avg       0.99      0.99      0.99     10225
   macro avg       0.99      0.99      0.99     10225
weighted avg       0.99      0.99      0.99     10225



#### Using random forest with hyperprameter tuning the f1-score on the X_test is 0.99 

In [47]:
# Random Forest variable importance using the best parameters obtained using gridsearch.

# importing modules
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier

#making the instance
model=RandomForestClassifier(n_estimators=300,min_samples_leaf=5,min_samples_split=5,criterion='entropy')

#Learning
model.fit(X_train, y_train)

#Prediction
prediction=model.predict(X_test)

# variable importance

feature_list = upsampled.columns
# Get numerical feature importances
importances = list(model.feature_importances_)# List of tuples with variable and importance
feature_importances = [(feature, round(importance, 2)) for feature, importance in zip(feature_list, importances)]# Sort the feature importances by most important first
feature_importances = sorted(feature_importances, key = lambda x: x[1], reverse = True)# Print out the feature and importances 
[print('Variable: {:20} Importance: {}'.format(*pair)) for pair in feature_importances];


Variable: x37                  Importance: 0.04
Variable: x75                  Importance: 0.04
Variable: x41                  Importance: 0.03
Variable: x58                  Importance: 0.03
Variable: x97                  Importance: 0.03
Variable: x1                   Importance: 0.02
Variable: x21                  Importance: 0.02
Variable: x40                  Importance: 0.02
Variable: x44                  Importance: 0.02
Variable: x51                  Importance: 0.02
Variable: x53                  Importance: 0.02
Variable: x63                  Importance: 0.02
Variable: x66                  Importance: 0.02
Variable: x70                  Importance: 0.02
Variable: x72                  Importance: 0.02
Variable: x83                  Importance: 0.02
Variable: x96                  Importance: 0.02
Variable: x99                  Importance: 0.02
Variable: x0                   Importance: 0.01
Variable: x2                   Importance: 0.01
Variable: x3                   Importanc

In [33]:
# test data

test.head()

x0         x1         x2         x3         x4         x5  \
0  -23.230884  -1.809757  12.380690  -4.101199 -60.760749 -22.957453   
1  138.561415   1.107473 -19.781009 -17.584757 -76.922109  71.816830   
2   -9.243047 -10.207303  -7.507803   3.152110 -14.915059  30.576213   
3    8.963713  17.580528  13.884170 -17.164185 -33.547539  19.288187   
4   27.431028  -6.232849  52.780835  -7.053903   5.679193 -29.618139   

         x6        x7        x8        x9    ...             x90       x91  \
0 -1.964078 -0.631029 -4.306616 -4.694198    ...     -200.759241  1.474824   
1 -0.418432  1.403957 -5.367051  0.039857    ...     -323.789253  5.539663   
2 -0.378178  2.606353  1.581679  4.802973    ...       27.816461 -0.977431   
3 -1.219019  5.574611 -3.879657  2.693111    ...     -374.983643 -2.631231   
4  1.331833  4.425692 -4.213258 -0.398755    ...     -352.113759  1.759678   

         x92   x93       x94        x95        x96        x97       x98  \
0  -8.057325  asia  3.418412 -10.364264 -18.055341  -6.236096 -4.909211   
1  -1.958140  asia -5.208792 -52.514599  -0.946514 -20.699992 -3.704683   
2  10.324237  asia  7.939948  19.031967   7.961947   6.761312 -1.072949   
3   2.206816  asia -8.234822 -14.077963 -11.476360  15.808888  6.361906   
4  -7.117473  asia  5.998549 -32.618465 -11.656808 -19.310773 -2.407052   

         x99  
0  47.679903  
1 -35.404859  
2  36.075623  
3  47.278303  
4 -54.988928  

[5 rows x 100 columns]

In [34]:
# convert test data to arry for prediction
test = test_df.values

In [35]:
# predict probabilities and save to output

proba = model.predict_proba(test)
one_class =[]
for i in range(len(proba)):
    one_class.append(proba[i][1])
one_class = pd.DataFrame({'one_class':one_class})
one_class.to_excel("output.xlsx")

#### Neural network model

In [36]:
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score
from keras.models import Sequential
from keras.layers import Dense
def build_classifier():
    classifier = Sequential()
    classifier.add(Dense(units = 60, kernel_initializer = 'uniform', activation = 'relu', input_dim = 129))
    classifier.add(Dense(units = 30, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))
    classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier
classifier = KerasClassifier(build_fn = build_classifier, batch_size = 1000, epochs = 100)
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 5, n_jobs = -1)
mean = accuracies.mean()
variance = accuracies.std()

# Improving the ANN
# Dropout Regularization to reduce overfitting if needed

# Tuning the ANN
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import GridSearchCV
from keras.models import Sequential
from keras.layers import Dense
def build_classifier(optimizer):
    classifier = Sequential()
    classifier.add(Dense(units = 60, kernel_initializer = 'uniform', activation = 'relu', input_dim = 129))
   # classifier.add(Dense(units = 30, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))
    classifier.compile(optimizer = optimizer, loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier
classifier = KerasClassifier(build_fn = build_classifier)
parameters = {'batch_size': [1000],
              'epochs': [100],
              'optimizer': ['adam']}
grid_search = GridSearchCV(estimator = classifier,
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 5)
grid_search = grid_search.fit(X_train, y_train)
best_parameters = grid_search.best_params_
best_accuracy = grid_search.best_score_


Using TensorFlow backend.


Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Use tf.cast instead.
Epoch 1/100
32717/32717 [==============================] - 2s 51us/step - loss: 0.6309 - acc: 0.7046
Epoch 2/100
32717/32717 [==============================] - 1s 40us/step - loss: 0.3613 - acc: 0.8435
Epoch 3/100
32717/32717 [==============================] - 1s 38us/step - loss: 0.3010 - acc: 0.8774
Epoch 4/100
32717/32717 [==============================] - 1s 39us/step - loss: 0.2509 - acc: 0.9046
Epoch 5/100
32717/32717 [==============================] - 1s 38us/step - loss: 0.2080 - acc: 0.9280
Epoch 6/100
32717/32717 [==============================] - 1s 39us/step - loss: 0.1780 - acc: 0.9428
Epoch 7/100
32717/32717 [==============================] - 1s 39us/step - loss: 0.1562 - acc: 0.9509
Epoch 8/100
32717/32717 [==============================] - 1s 43us/step - loss: 0.1398 - acc: 0.9578
Epoch 9/100
32717/32717 [==============================] - 1s 40us/step 

32717/32717 [==============================] - 1s 39us/step - loss: 0.0142 - acc: 0.9977
Epoch 76/100
32717/32717 [==============================] - 1s 41us/step - loss: 0.0137 - acc: 0.9977
Epoch 77/100
32717/32717 [==============================] - 1s 41us/step - loss: 0.0128 - acc: 0.9976
Epoch 78/100
32717/32717 [==============================] - 1s 38us/step - loss: 0.0124 - acc: 0.9983
Epoch 79/100
32717/32717 [==============================] - 1s 38us/step - loss: 0.0114 - acc: 0.9984
Epoch 80/100
32717/32717 [==============================] - 1s 38us/step - loss: 0.0110 - acc: 0.9985
Epoch 81/100
32717/32717 [==============================] - 1s 39us/step - loss: 0.0110 - acc: 0.9987
Epoch 82/100
32717/32717 [==============================] - 1s 38us/step - loss: 0.0118 - acc: 0.9979
Epoch 83/100
32717/32717 [==============================] - 1s 38us/step - loss: 0.0108 - acc: 0.9984
Epoch 84/100
32717/32717 [==============================] - 1s 38us/step - loss: 0.0104 - acc: 

32717/32717 [==============================] - 1s 38us/step - loss: 0.0223 - acc: 0.9956
Epoch 55/100
32717/32717 [==============================] - 1s 40us/step - loss: 0.0207 - acc: 0.9964
Epoch 56/100
32717/32717 [==============================] - 1s 40us/step - loss: 0.0198 - acc: 0.9965
Epoch 57/100
32717/32717 [==============================] - 1s 39us/step - loss: 0.0204 - acc: 0.9963
Epoch 58/100
32717/32717 [==============================] - 1s 38us/step - loss: 0.0201 - acc: 0.9961
Epoch 59/100
32717/32717 [==============================] - 1s 41us/step - loss: 0.0185 - acc: 0.9968
Epoch 60/100
32717/32717 [==============================] - 1s 39us/step - loss: 0.0176 - acc: 0.9969
Epoch 61/100
32717/32717 [==============================] - 1s 39us/step - loss: 0.0169 - acc: 0.9975
Epoch 62/100
32717/32717 [==============================] - 1s 39us/step - loss: 0.0174 - acc: 0.9972
Epoch 63/100
32717/32717 [==============================] - 1s 39us/step - loss: 0.0173 - acc: 

32718/32718 [==============================] - 1s 38us/step - loss: 0.0355 - acc: 0.9928
Epoch 35/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0345 - acc: 0.9929
Epoch 36/100
32718/32718 [==============================] - ETA: 0s - loss: 0.0331 - acc: 0.993 - 1s 38us/step - loss: 0.0331 - acc: 0.9934
Epoch 37/100
32718/32718 [==============================] - 1s 37us/step - loss: 0.0319 - acc: 0.9941
Epoch 38/100
32718/32718 [==============================] - 1s 41us/step - loss: 0.0312 - acc: 0.9938
Epoch 39/100
32718/32718 [==============================] - 1s 41us/step - loss: 0.0294 - acc: 0.9945
Epoch 40/100
32718/32718 [==============================] - 1s 37us/step - loss: 0.0291 - acc: 0.9946
Epoch 41/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0298 - acc: 0.9937
Epoch 42/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0267 - acc: 0.9953
Epoch 43/100
32718/32718 [==============================]

32718/32718 [==============================] - 1s 37us/step - loss: 0.0778 - acc: 0.9802
Epoch 14/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0735 - acc: 0.9807
Epoch 15/100
32718/32718 [==============================] - 1s 40us/step - loss: 0.0685 - acc: 0.9829
Epoch 16/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0644 - acc: 0.9840
Epoch 17/100
32718/32718 [==============================] - 1s 43us/step - loss: 0.0604 - acc: 0.9855
Epoch 18/100
32718/32718 [==============================] - 2s 47us/step - loss: 0.0581 - acc: 0.9860
Epoch 19/100
32718/32718 [==============================] - 1s 45us/step - loss: 0.0549 - acc: 0.9871
Epoch 20/100
32718/32718 [==============================] - 1s 45us/step - loss: 0.0526 - acc: 0.9876
Epoch 21/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0498 - acc: 0.9877
Epoch 22/100
32718/32718 [==============================] - 1s 39us/step - loss: 0.0473 - acc: 

32718/32718 [==============================] - 1s 38us/step - loss: 0.0026 - acc: 0.9999
Epoch 94/100
32718/32718 [==============================] - 1s 39us/step - loss: 0.0025 - acc: 0.9999
Epoch 95/100
32718/32718 [==============================] - 1s 39us/step - loss: 0.0027 - acc: 0.9999
Epoch 96/100
32718/32718 [==============================] - 1s 39us/step - loss: 0.0026 - acc: 0.9999
Epoch 97/100
32718/32718 [==============================] - 1s 37us/step - loss: 0.0023 - acc: 0.9999
Epoch 98/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0022 - acc: 1.0000
Epoch 99/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0023 - acc: 1.0000
Epoch 100/100
32718/32718 [==============================] - 1s 37us/step - loss: 0.0020 - acc: 1.0000
Epoch 1/100
32718/32718 [==============================] - 2s 57us/step - loss: 0.8006 - acc: 0.6714
Epoch 2/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.3801 - acc: 0

32718/32718 [==============================] - 1s 38us/step - loss: 0.0124 - acc: 0.9986
Epoch 74/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0113 - acc: 0.9985
Epoch 75/100
32718/32718 [==============================] - 1s 39us/step - loss: 0.0120 - acc: 0.9983
Epoch 76/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0129 - acc: 0.9979
Epoch 77/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0118 - acc: 0.9980
Epoch 78/100
32718/32718 [==============================] - 1s 38us/step - loss: 0.0113 - acc: 0.9985
Epoch 79/100
32718/32718 [==============================] - 1s 37us/step - loss: 0.0101 - acc: 0.9988
Epoch 80/100
32718/32718 [==============================] - 1s 42us/step - loss: 0.0105 - acc: 0.9985
Epoch 81/100
32718/32718 [==============================] - 1s 41us/step - loss: 0.0099 - acc: 0.9986
Epoch 82/100
32718/32718 [==============================] - 1s 39us/step - loss: 0.0093 - acc: 

40897/40897 [==============================] - 2s 40us/step - loss: 0.0147 - acc: 0.9978
Epoch 54/100
40897/40897 [==============================] - 2s 40us/step - loss: 0.0148 - acc: 0.9975
Epoch 55/100
40897/40897 [==============================] - 2s 40us/step - loss: 0.0140 - acc: 0.9976
Epoch 56/100
40897/40897 [==============================] - 2s 40us/step - loss: 0.0132 - acc: 0.9978
Epoch 57/100
40897/40897 [==============================] - 2s 41us/step - loss: 0.0125 - acc: 0.9982
Epoch 58/100
40897/40897 [==============================] - 2s 40us/step - loss: 0.0127 - acc: 0.9978
Epoch 59/100
40897/40897 [==============================] - 2s 40us/step - loss: 0.0151 - acc: 0.9968
Epoch 60/100
40897/40897 [==============================] - 2s 40us/step - loss: 0.0132 - acc: 0.9975
Epoch 61/100
40897/40897 [==============================] - 2s 39us/step - loss: 0.0114 - acc: 0.9985
Epoch 62/100
40897/40897 [==============================] - 2s 40us/step - loss: 0.0113 - acc: 

In [38]:
# predict probabilities and save to output

proba = grid_search.predict_proba(test)
one_class2 =[]
for i in range(len(proba)):
    one_class2.append(proba[i][1])
one_class2 = pd.DataFrame({'one_class':one_class2})
one_class2.to_excel("output2.xlsx")

In [39]:
best_parameters

{'batch_size': 1000, 'epochs': 100, 'optimizer': 'adam'}

In [46]:
print( 'Best accuracy using Neural network:',best_accuracy)

Best accuracy using Neural network: 0.987798616035406
